modules/dreamview/backend/data_collection_monitor/data_collection_monitor.cc

```diff
@@ -124,23 +125,66 @@ void DataCollectionMonitor::LoadConfiguration() {
       << "Unable to parse data collection configuration from file "
       << data_collection_config_path;
 
-  for (const auto& scenario_iter : data_collection_table_.scenario()) {
+  ConstructCategories();
+  for (const auto& scenario_iter : scenario_to_categories_) {
     const std::string& scenario_name = scenario_iter.first;
-    const Scenario& scenario = scenario_iter.second;
+    const auto& categories = scenario_iter.second;
 
-    for (const auto& category_iter : scenario.category()) {
+    for (const auto& category_iter : categories) {
       const std::string& category_name = category_iter.first;
-      const Category& category = category_iter.second;
-
       category_consecutive_frame_count_[scenario_name][category_name] = 0.0;
       category_frame_count_[scenario_name][category_name] = 0.0;
-      current_progress_json_[scenario_name][category.description()] = 0.0;
+      current_progress_json_[scenario_name][category_name] = 0.0;
     }
   }
 
   ADEBUG << "Configuration loaded.";
 }
 
+void DataCollectionMonitor::ConstructCategories() {
+  scenario_to_categories_.clear();
+
+  for (const auto& scenario_iter : data_collection_table_.scenario()) {
+    const std::string& scenario_name = scenario_iter.first;
+    const Scenario& scenario = scenario_iter.second;
+
+    Category category;
+    ConstructCategoriesHelper(scenario, 0, "", category,
+                              &scenario_to_categories_[scenario_name]);
+  }
+}
+
+void DataCollectionMonitor::ConstructCategoriesHelper(
+    const Scenario& scenario, int feature_idx,
+    std::string current_category_name, const Category& current_category,
+    std::unordered_map<std::string, Category>* categories) {
+  if (feature_idx == scenario.feature_size()) {
+    categories->insert({current_category_name, current_category});
+    return;
+  }
+
+  const Feature& feature = scenario.feature(feature_idx);
+  for (const auto& range : feature.range()) {
+    Category new_category(current_category);
+    new_category.push_back(range);
+
+    // set new category name by appending it with the current range name
+    std::stringstream new_category_name;
+    new_category_name << current_category_name;
+    const std::string& range_name = range.name();
+    if (feature.range().size() > 1 && !range_name.empty()) {
+      if (!current_category_name.empty()) {
+        new_category_name << ", ";
```

Comment:
I think we can use += to append to the category name, should be more efficient than <<